app/models/geo_entity.rb

```diff
@@ -32,19 +34,58 @@ def count_species
     Species.count_species(all_species)
   end
 
-  def occurrences
-    _occurrences = countries_geo_entity_stats.joins(:habitat).select('habitats.name, occurrence')
-    _occurrences = _occurrences.map(&:attributes).map { |attrs| attrs.slice('name', 'occurrence') }
+  def occurrences(real_values = false)
+    # Because of possible nil values for geo entity stat attributes, we can't just go by the occurrence
+    # value; nil values for a particular habitat will translate to data deficient even if confirmed present
+    # as there are two definitions:
+    #
+    # 1) We don't know if the habitat is present at all - this is the 'standard'
+    # data deficient definition
+    # 2) We know the habitat is present, we just don't know how much of it there is
+    occurrences = real_values ? determine_original_occurrences : determine_correct_occurrences
+
     hash = {}
 
-    _occurrences.map do |occurrence|
-      _occurrence = occurrence['occurrence']
+    occurrences.map do |occurrence|
       next if hash[occurrence['name']] == 'present'
-      hash[occurrence['name']] = _occurrence unless _occurrence == 'unknown'
+
+      # unless occurrence['occurrence'] == 'unknown'
+      hash[occurrence['name']] = occurrence['occurrence'] 
     end
+
     GeoEntityStat::BASE_OCCURRENCES.merge(hash)
   end
 
+  def fetch_needed_occurrence_attrs
+    countries_geo_entity_stats.joins(:habitat).select(
+      'habitats.name, 
+      occurrence, 
+      protected_value, 
+      total_value, 
+      geo_entity_stats.protected_percentage'
+    )
+  end
```

Comment:
Can this go below the `determine` methods so it reads high level functions to low level?

Reply:
Done, placed within private methods